### PR TITLE
Add connection check at startup when out_forward.

### PR DESCRIPTION
**Pull request by bootjp**
Added an option to check the out_forward node and communication when starting fluentd.

It is a matter with the following mailing list.
https://groups.google.com/forum/#!topic/fluentd/2ITnHbveYIM

Hello

I am using fluentd v1.2.6 in TLS encryption mode when forwarding in out_forward -> in_forward.

At fluentd startup I'd like to check if the out_forward process can connection with the in_forward node before listening for the forward port.
Does anyone know a good way about this?

The use case is below.
Existing applications are built across AWS and GCP, and the analysis base is in GCP.
I use out_forward TCP encryption for AWS and GCP connections.

The out_forward on the AWS side is redundant with the Active / Standby configuration.
When fluentd is deployed Switch from Active to Standby and return to the original Active / Standby when deployment is completed.

I think that it is desirable that fluentd gets an error as it is when the out_forward -> in_forward connection can not be performed by the new deployment so that the crash occurs.

The assumed error is a certificate validation failure or SHARD key mismatch.

I recognize that these errors are not checked at startup in fluentd now
See you.

I am not familiar with Ruby, and since I read the Fluentd code for the first time this time, please let me know if there is a better implementation.

This pull request is the same as below. #2180
I seemed to be merged because I was force pusn by the bot set in my account and the difference disappeared.
Bot has already been lifted.

### CODE REVIEW

File: test/plugin/test_out_forward.rb [hunk 1]
```diff
@@ -217,6 +217,18 @@ def read_ack_from_sock(sock, unpacker)
     assert_equal 2, d.instance.ack_response_timeout
   end
 
+  test 'verify_connection_at_startup is disabled in default' do
+    @d = d = create_driver(CONFIG)
+    assert_equal false, d.instance.verify_connection_at_startup
```
Comment (repeatedly) [inline test/plugin/test_out_forward.rb]: `assert_false` is better.

File: test/plugin/test_out_forward.rb [hunk 2]
```diff
@@ -217,6 +217,18 @@ def read_ack_from_sock(sock, unpacker)
...
+  end
+
+  test 'verify_connection_at_startup can be enabled' do
+    @d = d = create_driver(CONFIG + %[
+      verify_connection_at_startup true
+    ])
+    assert_equal true, d.instance.verify_connection_at_startup
```
Comment (repeatedly) [inline test/plugin/test_out_forward.rb]: `assert_true` is better

File: test/plugin/test_out_forward.rb [hunk 3]
```diff
@@ -774,4 +786,112 @@ def create_target_input_driver(response_stub: nil, disconnect: false, conf: TARG
       i.configure(conf)
     end
   end
+
+  test 'verify_connection_at_startup_nodes_are_not_available' do
```
Comment (repeatedly) [inline test/plugin/test_out_forward.rb]: You can use `sub_test_case` for grouping related tests.

```
sub_test_case 'verify_connection_at_startup' do
  test 'nodes are not available' do
    # test body
  end
end
```

File: lib/fluent/plugin/out_forward.rb [hunk 4]
```diff
@@ -253,6 +256,17 @@ def start
         @sock_ack_waiting = []
         thread_create(:out_forward_receiving_ack, &method(:ack_reader))
       end
+
+      if @verify_connection_at_startup
+        @nodes.each do |node|
+          begin
+            node.verify_connection
+          rescue StandardError => e
+            log.fatal e.message
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: log.fatal "forward's connection setting error: #{e.message}"

**Comment by bootjp** [comment]
Thank you for code review.
Several Ruby knowledge was deepened and it was very meaningful.

**Comment by repeatedly** [comment]
Merged. Thanks!

**Comment by bootjp** [comment]
@repeatedly 
Thanks for merge.
When will this change be released?

**Pull request merged**